### PR TITLE
DDPB-3525: Created a function to unarchive client from admin

### PR DESCRIPTION
**Pull request by MiaGordon91**
## Purpose
To allow high level admin users to unarchive a client for a deputy. A client may have been unarchived incorrectly or needs to be reported on by another deputy.

Fixes DDPB-3525

## Checklist
- [x] I have performed a self-review of my own code
- [ ] I have updated documentation (Confluence/ADR/tech debt doc) where relevant
- [x] I have added tests to prove my work
- [ ] The product team have approved these changes

### CODE REVIEW

File: api/tests/Behat/bootstrap/v2/ClientManagement/ClientManagementTrait.php [hunk 1]
```diff
@@ -369,4 +369,58 @@ public function theClientDoesNotHaveANamedDeputyAssociatedWithThem()
...
+            // This step is used as part of testing the unarchive button isnt here so swallow errors and assert on following step
+        }
+    }
+
+    /**
+     * @When an org deputy has an archived client
+     */
+    public function theDeputyHasAnArchivedClient()
+    {
+        $clientId = $this->interactingWithUserDetails->getClientId();
```
Comment (acsauk) [inline api/tests/Behat/bootstrap/v2/ClientManagement/ClientManagementTrait.php:394]: To make the use of this step more friendly we probably want to call `$this->assertInteractingWithUserIsSet();` here too so any future users of the code will get a useful failure error is the user details aren't set.

File: api/tests/Behat/bootstrap/v2/ClientManagement/ClientManagementTrait.php [hunk 2]
```diff
@@ -369,4 +369,58 @@ public function theClientDoesNotHaveANamedDeputyAssociatedWithThem()
...
+        $this->em->flush();
+    }
+
+    /**
+     * @Then the client should be unarchived
+     */
+    public function theClientShouldBeUnarchived()
+    {
+        $this->assertInteractingWithUserIsSet();
+        $this->iAmOnAdminClientDetailsPage();
```
Comment (acsauk) [inline api/tests/Behat/bootstrap/v2/ClientManagement/ClientManagementTrait.php:411]: Are we actually asserting on anything here? If we're displaying a flash message or something similar we could assert on that instead.

File: api/tests/Behat/features-v2/client-management/admin/unarchive.feature [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    Scenario: A super admin user can unarchived a client
+        Given a Professional Team Deputy has not started a health and welfare report
+        Given a super admin user accesses the admin app
+        And an org deputy has an archived client
+        When I visit the admin client archived page
+        And I attempt to unarchived the client
+        Then the client should be unarchived
+
+    @admin-manager @prof-team-hw-not-started
+    Scenario: An admin manager user can unarchived a client
```
Comment (acsauk) [inline api/tests/Behat/features-v2/client-management/admin/unarchive.feature:14]:
```suggestion
    Scenario: An admin manager user can unarchive a client
```

File: api/tests/Behat/features-v2/client-management/admin/unarchive.feature [hunk 4]
```diff
@@ -0,0 +1,29 @@
+@v2 @v2_admin @admin-client-unarchive
+Feature: Admin - Client Unarchived
+
+    @super-admin @prof-team-hw-not-started
+    Scenario: A super admin user can unarchived a client
+        Given a Professional Team Deputy has not started a health and welfare report
+        Given a super admin user accesses the admin app
+        And an org deputy has an archived client
+        When I visit the admin client archived page
```
Comment (acsauk) [inline api/tests/Behat/features-v2/client-management/admin/unarchive.feature:9]: There could be a little more context around this line - something like:

```suggestion
        When I visit the admin client archived page for the user I'm interacting with
```

File: client/src/Service/Client/Internal/ClientApi.php [hunk 5]
```diff
@@ -21,6 +21,7 @@ class ClientApi
     private const GET_CLIENT_BY_ID = 'client/%s';
     private const DELETE_CLIENT_BY_ID = 'client/%s/delete';
     private const UPDATE_CLIENT = 'client/upsert';
+    private const UNARCHIVED_CLIENT = 'client/%s/unarchived';
```
Comment (acsauk) [inline client/src/Service/Client/Internal/ClientApi.php:24]: Should this be `UNARCHIVE_CLIENT` and `'client/%s/unarchive'`?

File: client/templates/Admin/Client/Client/details.html.twig [hunk 6]
```diff
@@ -12,204 +12,217 @@
...
+        {#                        {% if not client.isArchived %} #}
+        {#                            <a href="" role="button" data-module="govuk-button" #}
+        {#                               class="govuk-button govuk-button--secondary"> #}
+        {#                                Un-archive client #}
+        {#                            </a> #}
+        {#                        {% endif %} #}
+        {#                    {% endif %} #}
+        {#                </dd> #}
+        {#            </div> #}
+        {#        {% endif %} #}
```
Comment (acsauk) [inline client/templates/Admin/Client/Client/details.html.twig:43]: Looks like we can drop this now

**Review by acsauk — CHANGES_REQUESTED** [review]
Great job - this should save us a bunch of time doing manual DB updates 💪 

A few commented out sections that we can drop and some context/tense update for the behat but overall looking nearly ready!

File: api/src/Controller/ClientController.php [hunk 7]
```diff
@@ -198,4 +198,22 @@ public function deleteAction(Request $request, $id)
...
         return [];
     }
+
+    /**
+     * @Route("/{id}/unarchived", methods={"PUT"}, requirements={"id":"\d+"})
+     * @Security("is_granted('ROLE_ADMIN_MANAGER')")
+     */
+    public function unarchiveClientAction(int $id)
+    {
+        // if the client is not in the database null then return NotFoundHttpException
```
[the remark this answers is not in the document]
Comment (coderste) [reply api/src/Controller/ClientController.php:208]: Haha yeah that's my bad I was teaching Mia about writing pseudo code :D

File: api/tests/Behat/bootstrap/v2/ClientManagement/ClientManagementTrait.php (hunk 2 above)
Comment (acsauk) [reply api/tests/Behat/bootstrap/v2/ClientManagement/ClientManagementTrait.php:412]: Ah OK gotcha 👍

**Review by acsauk — DISMISSED** [review]
:shipit:

**Review by acsauk — APPROVED** [review]
:shipit:

**Pull request merged**